Improve STI documentation

stevepolitodesign opened the pull request:
### Motivation / Background

This Pull Request has been created because the current [STI Guides][] don't document how to override or disable the [inheritance_column][].

~~Additionally, it's not clear that you can set `inheritance_column = :disabled`. Initially, I had tried  `inheritance_column = nil` which had the same behavior, but is not as clear as the former.~~

Setting `inheritance_column = nil` disabled STI.

### Detail

Add new sections to [STI Guides][] demonstrating how to override or disable the [inheritance_column][].

Also improve error message when `ActiveRecord::SubclassNotFound` is raised.

[STI Guides]: https://guides.rubyonrails.org/association_basics.html#single-table-inheritance-sti
[inheritance_column]: https://api.rubyonrails.org/classes/ActiveRecord/ModelSchema.html#method-c-inheritance_column

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

stevepolitodesign committed:
Improve STI documentation

Add new sections to [STI Guides][] demonstrating how to override or
disable the [inheritance_column][].

Also improve error message when `ActiveRecord::SubclassNotFound` is
raised.

[STI Guides]: https://guides.rubyonrails.org/association_basics.html#single-table-inheritance-sti
[inheritance_column]: https://api.rubyonrails.org/classes/ActiveRecord/ModelSchema.html#method-c-inheritance_column

pangolingo commented on guides/source/association_basics.md:
```diff
@@ -2912,6 +2912,46 @@ will run a query like:
+### Overriding the inheritance column
+
+There may be cases (like when working with a legacy database) where you need to
+override the name of the inheritance column. This can be achieved by with the
```
```suggestion
override the name of the inheritance column. This can be achieved with the
```

pangolingo commented on guides/source/association_basics.md:
```diff
@@ -2912,6 +2912,46 @@ will run a query like:
+### Disabling the inheritance column
+
+There may be cases (like when working with a legacy database) where you need to
+disable Single Table Inheritance all together. Otherwise, you'll raise
```
```suggestion
disable Single Table Inheritance altogether. Otherwise, you'll raise
```

p8 commented:
I'm not sure how this works. I can't find `:disabled` in the sources, so I'm guessing the column is ignored if it can't be found.

What would happen if you already have a column named `:disabled`?

stevepolitodesign committed:
Replace `:disabled` with `nil`

If there's a column named `disabled`, we'll still raise
`ActiveRecord::SubclassNotFound`. Setting `inheritance_column` to `nil`
works though.

I had mistakenly thought `:disabled` was the correct way to disable STI,
since this is what is done in [some tests][].

```ruby
Schema: cogs[ id, type, disabled, created_at, updated_at ]

class Cog < ApplicationRecord
  self.inheritance_column = nil
end

Cog.create! type: "Sprocket", disabled: true
=> #<Cog type: "Sprocket", disabled: "t">
```

[some tests]: https://github.com/rails/rails/blob/a4751751bced209550f38f48bb58b64406d475a5/activerecord/test/models/post.rb#L231

stevepolitodesign commented:
> I'm not sure how this works. I can't find `:disabled` in the sources, so I'm guessing the column is ignored if it can't be found.
> 
> What would happen if you already have a column named `:disabled`?

@p8 thank you for your diligence here!

Your instincts were correct. If there's a column named `disabled`, we'll still raise `ActiveRecord::SubclassNotFound`. Setting `inheritance_column` to `nil` works though.

I had mistakenly thought `:disabled` was the correct way to disable STI,
since this is what is done in [some tests](https://github.com/rails/rails/blob/a4751751bced209550f38f48bb58b64406d475a5/activerecord/test/models/post.rb#L231).


Addressed here: ef299a565629e2f6167f8c53d2aee1cd0e10c26f
